Feature/zero config pg autoctl

DimCitus opened the pull request:
Implement support for `--monitor` option and PG_AUTOCTL_MONITOR environment variable in many commands that don't need to have a local node already created. Those commands are:

    pg_autoctl create formation
    pg_autoctl drop formation

    pg_autoctl get|set formation settings
    pg_autoctl get|set formation number-sync-standbys
    pg_autoctl get|set node replication-quorum
    pg_autoctl get|set node candidate-priority

    pg_autoctl show uri
    pg_autoctl show events
    pg_autoctl show state
    pg_autoctl show settings
    pg_autoctl show standby-names

    pg_autoctl perform failover
    pg_autoctl perform switchover
    pg_autoctl perform promotion

The `make cluster` is also changed to use the PG_AUTOCTL_MONITOR environment variable rather than the monitor's PGDATA in the interactive terminal/session.

Fixes #437.

JelteF reviewed (changes requested):
I played around with this a bit. Some issues:
1. The `--monitor` argument doesn't actually seem to work (setting `PG_AUTOCTL_MONITOR` does work):
```
pg_autoctl show state --monitor 'postgres://autoctl_node@localhost:5500/pg_auto_failover?sslmode=prefer'
```

2. When using `--pgdata` you still see this error in the output:
```
12:02:39 14800 ERROR Failed to get value for environment variable 'PG_AUTOCTL_MONITOR', which is unset
```
3. When not specifying either `--pgdata` or `--monitor` the final fatal log should include something about `--monitor` and `PG_AUTOCTL_MONITOR` too:
```
11:55:49 4723 ERROR Failed to get value for environment variable 'PG_AUTOCTL_MONITOR', which is unset
11:55:49 4723 ERROR Failed to get value for environment variable 'PGDATA', which is unset
11:55:49 4723 FATAL Failed to set PGDATA either from the environment or from --pgdata
```

JelteF commented on src/bin/pg_autoctl/cli_do_tmux.c:
```diff
@@ -635,7 +647,7 @@ tmux_pg_autoctl_create_postgres(PQExpBuffer script,
 
 	sformat(monitor, sizeof(monitor),
-			"$(%s show uri --pgdata %s/monitor --monitor)",
+			"$(%s show uri --pgdata %s/monitor --formation monitor)",
```
I think if we change the args like this we would need a minor release instead of a patch. Since it could be a breaking change for some automation.

JelteF replied:
Also, maybe we should disallow using `monitor` as an actual formation name.

DimCitus replied:
Agreed about the minor release, it's a breaking change, although a very minor one. In another PR coming soon we reserve the monitor name for a formation where we register the monitor node itself...

JelteF commented on src/bin/pg_autoctl/cli_formation.c:
```diff
@@ -350,10 +404,21 @@ keeper_cli_formation_create(int argc, char **argv)
 	FormationConfig config = formationOptions;
 	Monitor monitor = { 0 };
 
-	if (!monitor_init_from_pgsetup(&monitor, &config.pgSetup))
```
Use `cli_monitor_init_from_option_or_config` instead?

DimCitus replied:
Can't, we have a `FormationConfig config` structure instance, not a `KeeperConfig config` one...

JelteF commented on src/bin/pg_autoctl/cli_formation.c:
```diff
@@ -388,10 +453,21 @@ keeper_cli_formation_drop(int argc, char **argv)
 		exit(EXIT_CODE_BAD_ARGS);
 	}
 
-	if (!monitor_init_from_pgsetup(&monitor, &config.pgSetup))
```
Use `cli_monitor_init_from_option_or_config` instead?

DimCitus replied:
Same as before.

JelteF commented on src/bin/pg_autoctl/cli_perform.c:
```diff
@@ -190,13 +204,25 @@ cli_perform_failover_getopts(int argc, char **argv)
-		exit(EXIT_CODE_BAD_ARGS);
+		if (!IS_EMPTY_STRING_BUFFER(options.pgSetup.pgdata))
+		{
+			log_warn("Given --monitor URI, the --pgdata option is ignored");
```
Is this really what we want? Other options are hard error or make `--pgdata` win.

Also does the same hold for setting `PG_AUTOCTL_MONITOR` and then using `--pgdata`?

DimCitus replied:
The other commands also prefer using --monitor over --pgdata, I think we're good here?

JelteF commented on src/bin/pg_autoctl/cli_show.c:
```diff
@@ -963,41 +1019,82 @@ cli_show_uri(int argc, char **argv)
+
+		/* cook the sslmode=prefer bits */
+		ssl.active = true;
+		ssl.sslMode = SSL_MODE_PREFER;
```
Any chance we could get this from the `monitor_pguri`? i.e. if monitor uri has `sslmode=verify-full&sslrootcert=somecert.crt` we would use those options.

DimCitus replied:
Done now.

JelteF reviewed (changes requested):
Some more issues:
1. 
```
PG_AUTOCTL_MONITOR='postgres://autoctl_node@localhost:5500/pg_auto_failover?sslmode=require' pg_autoctl create formation --formation 'test' --kind pgsql
13:48:11 24106 ERROR BUG: keeper_config_set_pathnames_from_pgdata: empty pgdata
```

2.
```
pg_autoctl create formation --formation 'test' --kind pgsql
13:51:00 30894 ERROR BUG: keeper_config_set_pathnames_from_pgdata: empty pgdata
```
3.
```
pg_autoctl create formation --formation 'test' --kind pgsql --monitor 'postgres://autoctl_node@localhost:5500/pg_auto_failover?sslmode=require'
13:51:37 32484 ERROR Failed to get value for environment variable 'PGDATA', which is unset
13:51:37 32484 FATAL Failed to set PGDATA either from the environment or from --pgdata
```
4. drop formation doesn't say it can use `--monitor` in the error message if you supply neither --monitor or --pgdata.
5.
```
PG_AUTOCTL_MONITOR='postgres://autoctl_node@localhost:5500/pg_auto_failover?sslmode=require' pg_autoctl drop formation --formation 'test'
13:53:27 3635 ERROR Failed to get value for environment variable 'PGDATA', which is unset
13:53:27 3635 FATAL Failed to set PGDATA either from the environment or from --pgdata
```
6. `show standby-names --group 5` does not error when there's no group 5 (might be the case on master too)

DimCitus commented on src/bin/pg_autoctl/parsing.c:
```diff
@@ -685,6 +685,74 @@ buildPostgresURIfromPieces(URIParams *uriParams, char *pguri)
+		{
+			strlcpy(ssl->crlFile, val, sizeof(ssl->crlFile));
+		}
+		else if (streq(key, "sslcert"))
+		{
+			strlcpy(ssl->serverCert, val, sizeof(ssl->serverCert));
+		}
+		else if (streq(key, "sslkey"))
+		{
+			strlcpy(ssl->serverKey, val, sizeof(ssl->serverKey));
+		}
```
I see what you mean. But in case we would reuse the function `parse_pguri_ssl_settings` ini other parts of the code I think it makes sense to still correctly parse those connection string parameters. After all those parameters are listed in https://www.postgresql.org/docs/current/libpq-connect.html.

JelteF replied:
Ah you are totally right. My mistake, I forgot these are used for TLS user certs.

DimCitus commented:
> Some more issues:

I have changed the way the `--monitor` and `--pgdata` and `PG_AUTOCTL_MONITOR` and `PGDATA` are understood in many places except for the formation code path, that I had done first. Sorry about that. I have now updated the code to match our current expectations.

> 1. `show standby-names --group 5` does not error when there's no group 5 (might be the case on master too)

Yeah, so the server-side function was returning NULL, which we didn't test against in `parseSingleValueResult`. So I have added an explicit `PQgetisnull` check in there, and also changed the server-side function to report an error in the case where we find zero nodes.

JelteF reviewed (approved):
Works well now
